### PR TITLE
Support attributes in limited cases (And refactored for debbugging...)

### PR DESCRIPTION
**Pull request by yutopp**

### CODE REVIEW

File: yuareg.el [hunk 1]
```diff
@@ -1248,32 +1248,50 @@ For use on `electric-indent-functions'."
   (when (fboundp 'smie-prec2->grammar)
     (let ((bnfprec2
            (smie-bnf->prec2
-            '((decls (decls "type" decls) (decls "d-let" decls)
-                     (decls "and" decls) (decls ";;" decls)
+            '((decls (decls "type" decls)
```
Comment (yutopp) [inline yuareg.el:1251]: Basically, added newlines for readability.

File: yuareg.el [hunk 2]
```diff
@@ -1248,32 +1248,50 @@ For use on `electric-indent-functions'."
...
-                     (def-in-exp))
+                     (def-in-exp)
+                     (item-attr)
+                     )
               (def-in-exp (defs "in" exp))
-              (def (var "d=" exp) (id "d=" datatype) (id "d=" module))
+              (def (var "d=" exp)
+                   (id "d=" datatype)
+                   (id "d=" module))
+              (item-attr ("[@@" id "]")) ; TODO: fix id
```
Comment (yutopp) [inline yuareg.el]: TODO

File: yuareg.el [hunk 3]
```diff
@@ -1724,9 +1743,9 @@ Return values can be
      ;; Some infix operators get a precedence based on their prefix, so we
      ;; collapse them into a canonical representative.
      ;; See http://caml.inria.fr/pub/docs/manual-ocaml/expr.html.
-     ((memq (aref tok 0) '(?* ?/ ?% ?+ ?- ?@ ?^ ?= ?< ?> ?| ?& ?$))
+     ((member (aref tok 0) '(?* ?/ ?% ?+ ?- ?@ ?^ ?= ?< ?> ?| ?& ?$))
       (cond
-       ((member tok '("|" "||" "&" "&&" "<-" "->")) tok)
+       ((member tok '("|" "||" "&" "&&" "<-" "->" "@" "@@" "@@@")) tok)
```
Comment (yutopp) [inline yuareg.el:1745]: treat `@`, `@@` and `@@@` explicitly.

File: yuareg.el [hunk 4]
```diff
@@ -1760,6 +1780,8 @@ Return values can be
    ((and (eq kind :before) (yuareg-smie--monadic-rule token)))
    ((and (equal token "and") (smie-rule-parent-p "type"))
     0)
+   ((and (equal token "[") (smie-rule-next-p "@@"))
```
Comment (yutopp) [inline yuareg.el]: Now, a syntax begin from `[ @@` will be treated as [item-attribute](https://caml.inria.fr/pub/docs/manual-ocaml/extn.html#sec260)

**Comment by nomaddo** [comment]
LGTM!!!

default

**Pull request merged**